### PR TITLE
Feature/stream compression

### PR DESCRIPTION
**Pull request by sandorw**
I'd like to use this to both compress streams and save on network costs by compressing client side.

This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1240)

### CODE REVIEW

**Comment by gsheasby** [comment]
@sandorw: so that we can prioritise this, could you let us know how urgent the change is for your product, and if you've done performance testing, what kind of boost you observed?

Also, are you aware of issue #1150? I'm currently working on this, and I wonder if this might actually be the problem you're suffering from. At first look, it appears that this PR does not touch that particular inefficiency.

**Comment by sandorw** [comment]
This isn't critical, but it would be great to have in the next few weeks - it's been a WIP on my end for a while between initial performance tests and then getting all the code assembled. Sorry for the lack of warning before now - I've been working with @clockfort on this locally.

For the particular data I'm storing, tests showed that I get 10x compression with this LZ4 algorithm. Writes were ~10-20% slower with the compression overhead in rough testing, but reads were ~2.5-3x faster including decompression. For our use case, we're writing immutable blobs ranging in size up to ~500 MB, so our workflow is very read skewed. I can rerun some tests and gather more data if you'd like.

I'm semi aware of #1150 - happy to account for it here, but I didn't want to bundle any API breaks with this change. The loadStreamAsFile() codepath requires a lot of extra code, so I'd be happy to not have to support it (all the other code paths just wrap a stream in the (de)compressing stream, but writing to a file requires me to stream the decompressed output myself). Is that issue something you're actively working on?

File: atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4Streams.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+package com.palantir.common.compression;
+
+import net.jpountz.lz4.LZ4Compressor;
+import net.jpountz.lz4.LZ4Factory;
+import net.jpountz.lz4.LZ4SafeDecompressor;
+
+public class LZ4Streams {
+    static final LZ4Factory factory = LZ4Factory.fastestInstance();
+    static final LZ4Compressor compressor = factory.fastCompressor();
+    static final LZ4SafeDecompressor decompressor = LZ4Factory.fastestInstance().safeDecompressor();
```
Comment (schlosna) [inline atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4Streams.java]: nit: `factory.safeDecompressor();` although maybe we should change this to:

```
static final LZ4Compressor compressor;
static final LZ4SafeDecompressor decompressor;
static {
    LZ4Factory factory = LZ4Factory.fastestInstance();
    compressor = factory.fastCompressor();
    decompressor = factory.safeDecompressor();
}

File: atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4CompressingInputStream.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        delegate.close();
+    }
+
+    private void writeFrameHeader() throws IOException {
+        System.arraycopy(getLittleEndianInt(LZ4Streams.MAGIC_VALUE), 0, compressedBuffer, 0, 4);
+        byte[] frameDescriptorBytes = frameDescriptor.toByteArray();
+        System.arraycopy(frameDescriptorBytes, 0, compressedBuffer, 4, LZ4Streams.FRAME_DESCRIPTOR_SIZE);
+    }
+
+    private byte[] getLittleEndianInt(int value) throws IOException {
```
Comment (schlosna) [inline atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4CompressingInputStream.java]: Might want to simplify to:

```
ByteBuffer buffer = ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(value);
buffer.flip();
return buffer.array();
```

**Comment by sandorw** [comment]
I will probably rework this a bit to move the (de)compression logic to a separate library.

**Comment by gsheasby** [comment]
Sounds good to me, sticking a "work in progress" tag on for now, and will review the PR once the compression logic is in a library.

**Comment by sandorw** [comment]
Slight change on our side - the other possible user of the library isn't around at the moment, and this will shortly be blocking another change that would like to use a compressed stream. I'd like to put the compression logic in here, then plan to remove it later if the library turns out to be useful enough. I'll add some light logic around versioning the stream store to handle backwards compatibility if we make any future changes.

If that's acceptable, I'll break this into two commits to make the review easier (stream compression vs. stream store). I'll account for @schlosna's comments there.

**Comment by sandorw** [comment]
Compression code now in this PR: https://github.com/palantir/atlasdb/pull/1312

**Comment by gsheasby** [comment]
That sounds good @sandorw - I've stuck "do not merge" on this PR. I'll close it once you've got the second PR up.

**Comment by sandorw** [comment]
Closed in favor of https://github.com/palantir/atlasdb/pull/1357

**Pull request closed without merging**